cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1611,6 +1636,23 @@ Result<ValueDescr> ResolveDecimalDivisionOutput(KernelContext*,
       });
 }
 
+Result<ValueDescr> ResolveTemporalOutput(KernelContext*,
+                                         const std::vector<ValueDescr>& args) {
+  auto left_type = checked_cast<const TimestampType*>(args[0].type.get());
+  auto right_type = checked_cast<const TimestampType*>(args[1].type.get());
+  DCHECK_EQ(left_type->id(), right_type->id());
+
+  if ((left_type->timezone() == "" || right_type->timezone() == "") &&
+      left_type->timezone() != right_type->timezone()) {
+    RETURN_NOT_OK(
+        Status::Invalid("Subtraction of zoned and non-zoned times is ambivalent. (",
+                        left_type->timezone(), right_type->timezone(), ")."));
+  }
+
+  auto type = duration(std::max(left_type->unit(), right_type->unit()));
```

Comment:
nit, but as used, the units should always match right?

Reply:
It's used to resolve duration resolution in `subtract(timestamp, timestamp) -> duration` and the timestamps could be of any resolution. Am I missing something?

Reply:
Ah, I was thinking, since we only register kernels where the units match, technically this max is redundant, but it doesn't hurt to have it. (Though perhaps it would be safer to DCHECK or return an error if they don't match to avoid accidentally doing arithmetic on incompatible types.)

Reply:
(I think return types are computed post-implicit-casts, right? So they should match here.)

Reply:
Ah, yes of course! Fixed. Thanks :).